File: GO/Structures/QueueDynamic.go
package Structures

type NodeQueue struct {
    value int
    next  *NodeQueue
}

type QueueDynamic struct {
    front *NodeQueue
    rear  *NodeQueue
    size  int
}

func NewQueueDynamic() *QueueDynamic {
    return &QueueDynamic{
        front: nil,
        rear:  nil,
        size:  0,
    }
}

func (q *QueueDynamic) Enqueue(value int) {
    newNode := &NodeQueue{value: value, next: nil}
    
    if q.rear == nil {
        q.front = newNode
        q.rear = newNode
    } else {
        q.rear.next = newNode
        q.rear = newNode
    }
    q.size++
}

func (q *QueueDynamic) Dequeue() (int, bool) {
    if q.front == nil {
        return 0, false
    }
    
    value := q.front.value
    q.front = q.front.next
    
    if q.front == nil {
        q.rear = nil
    }
    
    q.size--
    return value, true
}

func (q *QueueDynamic) ToArray() []int {
    result := make([]int, q.size)
    current := q.front
    i := 0
    
    for current != nil {
        result[i] = current.value
        current = current.next
        i++
    }
    
    return result
}

func (q *QueueDynamic) FromArray(arr []int) {
    q.front = nil
    q.rear = nil
    q.size = 0
    
    for _, v := range arr {
        q.Enqueue(v)
    }
}

func (q *QueueDynamic) Size() int {
    return q.size
}

File: PYTHON/structures/queue_dynamic.py
# structures/queue_dynamic.py
from typing import List, Optional, Tuple
import sys
sys.path.append('..')
from utils.data_structure import DataStructure


class NodeQueue:
    """Nó para fila encadeada"""
    def __init__(self, value: int):
        self.value = value
        self.next: Optional['NodeQueue'] = None


class QueueDynamic(DataStructure):
    """Fila com alocação dinâmica (lista encadeada)"""
    
    def __init__(self):
        self.front: Optional[NodeQueue] = None
        self.rear: Optional[NodeQueue] = None
        self._size = 0
    
    def enqueue(self, value: int) -> None:
        new_node = NodeQueue(value)
        
        if self.rear is None:
            self.front = new_node
            self.rear = new_node
        else:
            self.rear.next = new_node
            self.rear = new_node
        
        self._size += 1
    
    def dequeue(self) -> Tuple[int, bool]:
        if self.front is None:
            return 0, False
        
        value = self.front.value
        self.front = self.front.next
        
        if self.front is None:
            self.rear = None
        
        self._size -= 1
        return value, True
    
    def to_array(self) -> List[int]:
        result = []
        current = self.front
        
        while current is not None:
            result.append(current.value)
            current = current.next
        
        return result
    
    def from_array(self, arr: List[int]) -> None:
        self.front = None
        self.rear = None
        self._size = 0
        
        for v in arr:
            self.enqueue(v)
    
    def size(self) -> int:
        return self._size